src/client/Microsoft.Identity.Client/Extensibility/CdtAuthenticationScheme.cs

```diff
@@ -0,0 +1,296 @@
+        public const string CdtEncryptedValue = "A256CBC-HS256";
+        public const string CdtRequestConfirmation = "req_ds_cnf";
+
+        private readonly CdtCryptoProvider _cdtCryptoProvider;
+        private readonly string _constraints;
+        private readonly string _dsReqCnf;
+
+        /// <summary>
+        /// Creates Cdt tokens, i.e. tokens that are bound to an HTTP request and are digitally signed.
+        /// </summary>
+        /// <remarks>
+        /// Currently the signing credential algorithm is hard-coded to RSA with SHA256. Extensibility should be done
+        /// by integrating Wilson's SigningCredentials
+        /// </remarks>
+        public CdtAuthenticationScheme(string constraints, X509Certificate2 certificate)
+        {
+            _constraints = constraints ?? throw new ArgumentNullException(nameof(constraints));
+
+            _cdtCryptoProvider = new CdtCryptoProvider(certificate);
+
+            _dsReqCnf = _cdtCryptoProvider.CannonicalPublicKeyJwk;
+        }
+
+        public TokenType TelemetryTokenType => TokenType.Bearer;
+
+        public string AuthorizationHeaderPrefix => Constants.BearerAuthHeaderPrefix;
+
+        public string AccessTokenType => Constants.BearerAuthHeaderPrefix;
+
+        /// <summary>
+        /// For Cdt, we chose to use the base64(jwk_thumbprint)
+        /// </summary>
+        public string KeyId { get; }
+
+        public IReadOnlyDictionary<string, string> GetTokenRequestParams()
+        {
+            var temp = "eyJrdHkiOiAiUlNBIiwgIm4iOiAiMUNMeDNXRW1NWlQ3el92Szc2ZHBaVVNwX2kyMEEza0Y0OWVpemtCQTBFSTJ4el9pcldTcm9BamJrRTk4dlp3SFM0QVlQV2I5WEd2YTFPYVNMX0RqQTFPTG1nSll4Uk45cU5jd1lKeGhsN3hqaGJlU25RMUMtR1NNS3ZWRzJnaDdQUlhaaU1xVXFuOWt3UzBXa1RoNDhSREMxR0xhTFFfNzZmb0dZMmo0MlNvel9XYnNRemtnVGo0TDVaVTZTWjJ3QTFwMlZ6WFliOVd1M3A4U2VuV3JCTDUzOWhUZjVGelp0b1E0R2IxNzMzVzFmWVFsUkotYUZVMTFfdEc1Umx2Ui1nSWFweHJMWkFKM1NHM28wQ2ZPa2FaejdKT2RETnJHNnE4akF3ZmJOdFJ1eDYzbnJZZ0FHc3VhemlXalZxRnZiclNMX2Mya3dZaDlZUl9uYVJFOG1RPT0iLCAiZSI6ICJBUUFCIn0%3D";
+            return new Dictionary<string, string>() {
+                { OAuth2Parameter.TokenType, Constants.BearerAuthHeaderPrefix},
+                { CdtRequestConfirmation, Base64UrlHelpers.Encode(_dsReqCnf)}
+            };
+        }
+
+        public void FormatResult(AuthenticationResult authenticationResult)
+        {
+            var header = new JObject();
+            header[JsonWebTokenConstants.Algorithm] = NoAlgorythmPrefix;
+            header[JsonWebTokenConstants.Type] = CdtTokenType;
+
+            //TODO: determine what happens if nonce is not present
+            authenticationResult.AdditionalResponseParameters.TryGetValue(CdtNonce, out string nonce);
+            var body = CreateCdtBody(authenticationResult.AccessToken, nonce);
+
+            string constraintToken = CreateJWS(JsonHelper.JsonObjectToString(body), JsonHelper.JsonObjectToString(header), false);
+            authenticationResult.AccessToken = constraintToken;
```

Comment:
publicly `authenticationResult.AccessToken` is `get` only, so this will not work as an extension.

Reply:
A setter was exposed as part of this PR: 

https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/4859/files/27f379189c9b6e21a01a7471c6625a0c0b6aec4b#diff-6e9f3460583c54b690e5d7d155cb337653c519dd7cd5f513cc36ff82c3417849R202